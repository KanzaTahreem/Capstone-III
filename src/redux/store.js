import { configureStore } from '@reduxjs/toolkit';
import detailsSlice from './detailsSlice';
import currenciesSlice from './currenciesSlice';
import globalSlice from './globalSlice';
import trendingSlice from './trendingSlice';
import chartSlice from './chartSlice';

const store = configureStore({
  reducer: {
    currency: currenciesSlice,
    global: globalSlice,
    trending: trendingSlice,
    details: detailsSlice,
    chart: chartSlice,
  },
});

export default store;
