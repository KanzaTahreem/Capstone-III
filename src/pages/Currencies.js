import React from 'react';
import millify from 'millify';
import { useSelector, useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { BiSearch } from 'react-icons/bi';
import { userInput } from '../redux/currenciesSlice';
import CurrencyData from '../components/CurrencyData';
import Loader from '../components/Loader';

const Currencies = () => {
  const { filteredSearchArray, loading, error } = useSelector((state) => state.currency);
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const handleSearch = (e) => {
    dispatch(userInput(e.target.value));
  };

  const openCurrencyDetails = (coin) => {
    navigate(`/currency/${coin.id}/details`, { state: { coin } });
  };

  if (loading) {
    return <Loader />;
  }

  if (error) {
    return <p>{error}</p>;
  }

  return (

    <section className="coins_container">
      <header>
        <div>
          <h2 className="details_title title">
            Top 100 Crypto Currencies Data
          </h2>
          <h3 className="details_subtitle subtitle">
            A comprehensive overview of the top 100 cryptocurrencies, including their
            {' '}
            market capitalization, price trends, trading volumes, and key features
          </h3>
        </div>
        <div>
          <form>
            <input type="text" name="name" id="name" placeholder="Search here..." onChange={handleSearch} data-testid="input" />
            <BiSearch />
          </form>
        </div>
      </header>
      <main>
        <div className="currency-list">
          <div className="labels">
            <p>#</p>
            <p>Coin</p>
            <p>Sym</p>
            <p>Price</p>
            <p>1h</p>
            <p>24h</p>
            <p>7d</p>
            <p>24h Vol</p>
            <p>Mkt Cap</p>
          </div>
          <ul className="currencies" data-testid="coins-list-container">
            {filteredSearchArray.map((coin) => (
              <button key={coin.id} type="button" onClick={() => openCurrencyDetails(coin)} className="currency_btn">
                <li className="currency_card">
                  <CurrencyData
                    rank={millify(coin?.market_cap_rank)}
                    image={coin?.image}
                    name={coin?.name}
                    symbol={coin?.symbol}
                    currentPrice={`$ ${millify(coin?.current_price)}`}
                    oneHourPriceChange={`${millify(coin?.price_change_percentage_1h_in_currency)} %`}
                    oneDayPriceChange={`${millify(coin?.price_change_percentage_24h_in_currency)} %`}
                    oneWeekPriceChange={`${millify(coin?.price_change_percentage_7d_in_currency)} %`}
                    totalVolume={`$ ${millify(coin?.total_volume)}`}
                    marketCap={`$ ${millify(coin?.market_cap)}`}
                  />
                </li>
              </button>
            ))}
          </ul>
        </div>
      </main>
    </section>
  );
};

export default Currencies;
