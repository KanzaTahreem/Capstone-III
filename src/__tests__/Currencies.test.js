import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/store';
import Currencies from '../pages/Currencies';

const currenciesArray = [
  {
    id: 'bitcoin',
    symbol: 'btc',
    name: 'Bitcoin',
    image: 'https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579',
  },
  {
    id: 'ethereum',
    symbol: 'eth',
    name: 'Ethereum',
    image: 'https://assets.coingecko.com/coins/images/279/large/ethereum.png?1595348880',
  }];

describe('Testing Currencies', () => {
  jest.spyOn(global, 'fetch').mockResolvedValue({
    json: jest.fn().mockResolvedValue(currenciesArray),
  });
  let currencies;
  beforeAll(async () => {
    currencies = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Currencies />
          </Router>
        </Provider>,
      )
      .toJSON();
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });
  it('should renders correctly', () => {
    expect(currencies).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(currencies).toMatchSnapshot();
  });

  it('renders the component', () => {
    waitFor(() => expect(screen.getByText('Top 100 Crypto Currencies Data')).toBeInTheDocument());
  });

  it('should have an input field', async () => {
    const inputEl = waitFor(() => screen.findByTestId('input'));
    waitFor(() => expect(inputEl).toBeInTheDocument());
    waitFor(() => expect(inputEl).toHaveAttribute('type', 'text'));
  });

  it('should render two currencies', async () => {
    const ul = waitFor(() => screen.findByTestId('coins-list-container'));
    waitFor(() => expect(ul.querySelectorAll('li').length).toBe(2));
  });
});
