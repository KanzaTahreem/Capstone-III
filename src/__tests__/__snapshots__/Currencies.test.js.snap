// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Currencies should match the snapshot 1`] = `
<section
  className="coins_container"
>
  <header>
    <div>
      <h2
        className="details_title title"
      >
        Top 100 Crypto Currencies Data
      </h2>
      <h3
        className="details_subtitle subtitle"
      >
        A comprehensive overview of the top 100 cryptocurrencies, including their
         
        market capitalization, price trends, trading volumes, and key features
      </h3>
    </div>
    <div>
      <form>
        <input
          data-testid="input"
          id="name"
          name="name"
          onChange={[Function]}
          placeholder="Search here..."
          type="text"
        />
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
          />
        </svg>
      </form>
    </div>
  </header>
  <main>
    <div
      className="currency-list"
    >
      <div
        className="labels"
      >
        <p>
          #
        </p>
        <p>
          Coin
        </p>
        <p>
          Sym
        </p>
        <p>
          Price
        </p>
        <p>
          1h
        </p>
        <p>
          24h
        </p>
        <p>
          7d
        </p>
        <p>
          24h Vol
        </p>
        <p>
          Mkt Cap
        </p>
      </div>
      <ul
        className="currencies"
        data-testid="coins-list-container"
      />
    </div>
  </main>
</section>
`;
